### DISCUSSION TITLE
Detect physical switch press on a smart relay (Sonoff Mini etc)

### DISCUSSION THREAD

**Discussion by Rapti** [Koenkk/zigbee2mqtt#23311] [Devices]
I have a few [Sonoff Mini](https://www.zigbee2mqtt.io/devices/ZBMINI.html#sonoff-zbmini) and [Sonoff Mini L2](https://www.zigbee2mqtt.io/devices/ZBMINIL2.html#sonoff-zbminil2) smart relays scattered behind several wall switches. In some rooms, when someone uses the physical wall switch, I'd like to trigger additional actions such as switching other lights in the room, but not when the relay is switched via Zigbee. Is there a way to somehow expose physical switch uses to MQTT? There is an option called `state_action` that I was hoping would achieve this, but apparently all it does is add an additional `action` property to the JSON Payload sent onto the MQTT topic, regardless of what triggered the switch.

Is this possible? Or would this have to be supported by the device's firmware?

**Comment by duncanhancox** [comment]
i havnt found a zigbee relay that allows the switch to be detached from the action, unlike a Shelly wifi switch or dimmer

my solution was to change the switch / relay to a cheap dual channel model, ch1 state_action sends the switch status to intermediate software (in my case, Loxone via mqtt) and ch2 is activated by the outcome of the intermediate software again via mqtt / z2m

**Reply by Rapti** [reply]
I'm aware that these things don't support detaching, but I hoped to still be able to detect switch presses.
Real decoupling would be really useful as well though! Sonoff's newer Wifi variants even support this, but it seems someone has yet to release a Zigbee relay with this feature.

**Reply by chris-1243** [reply]
Ubisys is offering such option but they are expensive...

**Comment by francisp2** [comment]
What home automation software are you using ?

For HA I use this template trigger
```
- trigger:
    - platform: state
      entity_id: light.lichtkeuken
      to: "on"
  sensor:
    - name: "Keuken Licht On Context"
      unique_id: '018e8402-759b-761b-bf24-272b807826e6'
      state: >
          {% set c_id = trigger.to_state.context.id %}
          {% set c_parent = trigger.to_state.context.parent_id %}
          {% set c_user = trigger.to_state.context.user_id %}
          {% if c_id != none and c_parent == none and c_user == none %}
            physical
          {% elif c_id != none and c_parent == none and c_user != none %}
            dashboard_ui
          {% elif c_id != none and c_parent != none and c_user == none %}
            automation
          {% else %}
            unknown
          {% endif %}
```

**Reply by Rapti** [reply]
Thanks, I use OpenHAB though. My current workaround is to just ignore updates coming from the relay for a few seconds after I've sent a command to it. Whenever I receive an update from the relay and I haven't recently sent a command to it, I'll assume the physical switch was used.

I believe both of our solutions will detect Zigbee bindings as physical switch uses, though.

**Comment by alexeiw123** [comment]
very similarly to @francisp2; (<-- Theirs is more flexible in that it gives you other options, mine is very specific to physical actions)

I use the `trigger.to_state.context` parts to determine physical device presses.

This condition will only allow an automation to run from a physical press.

```
condition:
  - condition: template
    value_template: "{{ trigger.to_state.context.id != none }}"
  - condition: template
    value_template: "{{ trigger.to_state.context.user_id == none }}"
  - condition: template
    value_template: "{{ trigger.to_state.context.parent_id == none }}"
```
